README.md
# object_path

> Create nested values and any intermediaries using list notation `['a','b',c']` OR dot notation `"a.b.c"`.

## Still in development, will be ready very soon (1 week max)

This is a `Dart` copy of the `js` projects [set-value](https://www.npmjs.com/package/set-value) by [Jon Schlinkert](https://github.com/jonschlinkert) and [object-path](https://www.npmjs.com/package/object-path) by [Mario Casciaro](https://github.com/mariocasciaro).
Please consider following either (or all) projects and giving some stars :star: to show your :heart: and support.
Also, feel free to make pull requests, open issues, or just let me know if I've messed up. The goal is to keep this project functionality in-line with the originals as far as possible. Bear in mind that some functionality is slightly different to accommodate for the typed nature of `Dart`, as well as my personal preference (if you want anything changed...my arm can be twisted).


Lastly, please consider following the project author [Ferdinand Steenkamp](https://github.com/Ferdzzzzzzzz/).

## Install
Install with [pub.dev](https://pub.dev/), add this to your `pubspec.yaml`:

    dependencies:
        object_path:
Then run `flutter pub get`.

## Usage

Import the package:
    
    import 'package:object_path/object_path.dart';

#### Note: I made strong distinctions between `list` and `dot`-notation to prevent confusion.

### Using `dot` notation:
    
    void main() {
        
        var mock = <String, dynamic>{}
        var objPath = ObjectPath();
        
        // set a value in a map
        mock = objPath.setDot(mock, 'a.b.c', 'Value', splitAt = '.', escapeWith = '\\');
        
        // unset value in a map
        mock = objPath.unsetDot(mock, 'a.b.c');
        
        // get value in a map, returns null if it doesn't exists
        print(objPath.getDot(mock, 'a.b.c'));

        // escape split character
        mock = objPath.setDot(mock, 'a.b.c\\.d', 'Value');
    }

Notes:

*Escaping:* You can escape a split character using `\\`, alternatively, you can add your own escape character using the optional `escapeWith` option.

*Dot Notation:* The dot notation obviously uses...dots. Alternatively, you can add your own dot character using the optional `splitAt` option.

### Using `List<String>` method:
    
    void main() {
        
        var mock = <String, dynamic>{}
        var objPath = ObjectPath();
        
        // set a value in a map
        mock = objPath.set(mock, ['a', 'b', 'c'], 'Value');
        
        // unset value in a map
        mock = objPath.unset(mock, ['a', 'b', 'c']);
        
        //get value in a map, returns null if it doesn't exists
        print(objPath.get(mock, ['a','b','c']));
    }



### About

Pull requests and issues always welcome. There is a lot more information on this project at the original [set-value](https://www.npmjs.com/package/set-value). 

### Author

**Ferdinand Steenkamp**

* [Website](https://ferdzz.com)
* [GitHub Profile](https://github.com/Ferdzzzzzzzz)
* [Twitter Profile](https://twitter.com/therealferdz)
* [LinkedIn Profile](https://www.linkedin.com/in/ferdinand-steenkamp-552335181/)



### License

Copyright © 2020, [Ferdinand Steenkamp](https://github.com/Ferdzzzzzzzz).
Released under the [BSD License](LICENSE).





CHANGELOG.md
## 0.0.1-dev.1

- Actively developing, not ready for use


test/object_path_test.dart
import 'package:object_path/object_path.dart';
import 'package:test/test.dart';

void main() {
  group('Test List Notation', () {
    test('Test set function', () {
      var objPath = ObjectPath();
      var testMap = <String, dynamic>{
        'Key 1': 'Val 1',
        'Key 2': 'Val 2',
        'Key 3': {'Key 3.1': 'Val 3'}
      };

      testMap =
          objPath.setList(testMap, ['Key 4', 'Key 4.1', 'Key 4.1.1'], 'Value 4');
      testMap = objPath.setList(testMap, ['Key 5'], [1, 2]);

      var assertTest = <String, dynamic>{
        'Key 1': 'Val 1',
        'Key 2': 'Val 2',
        'Key 3': {'Key 3.1': 'Val 3'},
        'Key 4': {
          'Key 4.1': {
            'Key 4.1.1': 'Value 4',
          }
        },
        'Key 5': [1, 2]
      };

      expect(testMap, equals(assertTest));
    });

    test('Test unset function', () {
      var objPath = ObjectPath();
      var testMap = <String, dynamic>{
        'Key 1': 'Val 1',
        'Key 2': 'Val 2',
        'Key 3': {'Key 3.1': 'Val 3'},
        'Key 5': {
          'Key 5.1': {
            'Key 5.1.2': [1, 2]
          },
          'Key 5.2': {
            'Key 5.2.1': 'Val 5.2.1',
            'Key 5.2.2': 'Val 5.2.2',
          }
        }
      };

      testMap = objPath.unsetList(testMap, ['Key 5', 'Key 5.2', 'Key 5.2.2']);
      testMap = objPath.unsetList(testMap, ['Key 1']);

      var assertTest = <String, dynamic>{
        'Key 2': 'Val 2',
        'Key 3': {'Key 3.1': 'Val 3'},
        'Key 5': {
          'Key 5.1': {
            'Key 5.1.2': [1, 2]
          },
          'Key 5.2': {
            'Key 5.2.1': 'Val 5.2.1',
          }
        }
      };

      expect(testMap, equals(assertTest));
    });
  });

  group('Test Dot Notation', () {
    test('Test setDot function', () {
      var objPath = ObjectPath();
      var testMap = <String, dynamic>{
        'Key 1': 'Val 1',
        'Key 2': 'Val 2',
        'Key 3': {'Key 3.1': 'Val 3'}
      };

      testMap =
          objPath.setDot(testMap, 'Key 4.Key 4\\.1.Key 4\\.1\\.1', 'Value 4');
      testMap = objPath.setDot(testMap, 'Key 5', [1, 2]);

      var assertTest = <String, dynamic>{
        'Key 1': 'Val 1',
        'Key 2': 'Val 2',
        'Key 3': {'Key 3.1': 'Val 3'},
        'Key 4': {
          'Key 4.1': {
            'Key 4.1.1': 'Value 4',
          }
        },
        'Key 5': [1, 2]
      };

      expect(testMap, equals(assertTest));
    });

    test('Test unsetDot function', () {
      var objPath = ObjectPath();
      var testMap = <String, dynamic>{
        'Key 1': 'Val 1',
        'Key 2': 'Val 2',
        'Key 3': {'Key 3.1': 'Val 3'},
        'Key 5': {
          'Key 5.1': {
            'Key 5.1.2': [1, 2]
          },
          'Key 5.2': {
            'Key 5.2.1': 'Val 5.2.1',
            'Key 5.2.2': 'Val 5.2.2',
          }
        }
      };

      testMap = objPath.unsetDot(testMap, 'Key 5.Key 5\\.2.Key 5\\.2\\.2');
      testMap = objPath.unsetDot(testMap, 'Key 1');

      var assertTest = <String, dynamic>{
        'Key 2': 'Val 2',
        'Key 3': {'Key 3.1': 'Val 3'},
        'Key 5': {
          'Key 5.1': {
            'Key 5.1.2': [1, 2]
          },
          'Key 5.2': {
            'Key 5.2.1': 'Val 5.2.1',
          }
        }
      };

      expect(testMap, equals(assertTest));
    });
  });
}


lib/src/object_path_base.dart
class ObjectPath {
  ObjectPath();

  ///
  /// set
  ///
  /// Takes a json object, path and value,
  /// returns the object with that value set
  /// at the path
  ///
  /// @Map<String,dynamic> target
  /// @List<String> path
  /// @dynamic value
  ///
  Map<String, dynamic> setList(
    Map<String, dynamic> target,
    List<String> path,
    dynamic value,
  ) {
    if (!(path is List<String>)) {
      return target;
    }

    if (path.length == 1) {
      _result(target, path[0], value);
      return target;
    }

    var orig = target;
    var len = path.length;

    for (var i = 0; i < len; i++) {
      var prop = path[i];

      if (!(target[prop] is Map<String, dynamic>)) {
        target[prop] = <String, dynamic>{};
      }

      if (i == len - 1) {
        _result(target, prop, value);
        break;
      }

      target = target[prop];
    }

    return orig;
  }

  ///
  /// unset
  ///
  /// Takes a json object and unsets a key
  ///
  /// @Map<String,dynamic> target
  /// @String path (Map key)
  ///
  Map<String, dynamic> unsetList(
    Map<String, dynamic> target,
    List<String> path,
  ) {
    if (!(path is List<String>)) {
      return target;
    }

    if (path.length == 1) {
      target.remove(path[0]);
      return target;
    }

    var orig = target;
    var len = path.length;

    for (var i = 0; i < len; i++) {
      var prop = path[i];

      if (!(target.containsKey(prop))) {
        break;
      }

      if (i == len - 1) {
        target.remove(path[i]);
        break;
      }

      target = target[prop];
    }

    return orig;
  }

  ///
  /// get
  ///
  /// Takes a json object and path and returns the value at that path
  ///
  /// @Map<String,dynamic> target
  /// @String path (Map key)
  ///
  ///
  dynamic getList(
    Map<String, dynamic> target,
    List<String> path,
    {
      dynamic defaultVal,
    }
  ) {
    if (!(path is List<String>)) {
      return target;
    }

    if (path.length == 1) {
      return target[path[0]];
    }

    var len = path.length;

    for (var i = 0; i < len; i++) {
      var prop = path[i];

      if (!(target.containsKey(prop))) {
        break;
      }

      if (i == len - 1) {
        return target[path[i]];
      }

      target = target[prop];
    }
    return defaultVal;
  }

  ///
  /// setDot
  /// Add a value to the Map key using dot-notation
  ///
  /// @Map<String,dynamic> target
  /// @String path (Map key)
  /// @dynamic value
  ///
  Map<String, dynamic> setDot(
    Map<String, dynamic> target,
    String path,
    dynamic value, {
    String splitAt = '.',
    String escapeWith = '\\',
  }) {
    return setList(target, _split(path, splitAt, escapeWith), value);
  }

  ///
  /// unsetDot
  /// Remove a key from the map using dot-notation
  ///
  /// @Map<String,dynamic> target
  /// @String path (Map key)
  /// @optional String splitAt
  /// @optional String escapeWith
  ///
  Map<String, dynamic> unsetDot(
    Map<String, dynamic> target,
    String path, {
    String splitAt = '.',
    String escapeWith = '\\',
  }) {
    return unsetList(target, _split(path, splitAt, escapeWith));
  }


  ///
  /// getDot
  /// Get a value at any path using dot-notation
  ///
  /// @Map<String,dynamic> target
  /// @String path (Map key)
  /// @optional String splitAt
  /// @optional String escapeWith
  ///
  dynamic getDot(
      Map<String, dynamic> target,
      String path, {
        String splitAt = '.',
        String escapeWith = '\\',
        dynamic defaultVal,
      }) {
    return getList(target, _split(path, splitAt, escapeWith));
  }

  ///
  /// _result
  /// Add a value to the Map key
  ///
  /// @Map<String,dynamic> target
  /// @String path (Map key)
  /// @dynamic value
  ///
  void _result(
    Map<String, dynamic> target,
    String path,
    dynamic value,
  ) {
    target[path] = value;
  }

  ///
  /// _split
  /// Split an String with dot-notation
  ///
  /// @Map<String,dynamic> target
  /// @String splitAt
  /// @String escapeWith
  ///

  List<String> _split(String path, String splitAt, String escapeWith) {
    var keys = path.split(splitAt);

    var res = <String>[];
    String prop;

    for (var i = 0; i < keys.length; i++) {
      prop = keys[i];
      var lastChar = prop.length - 1;
      while (prop.substring(lastChar) == escapeWith) {
        prop = prop.substring(0, lastChar) + splitAt + keys[++i];
        lastChar = prop.length - 1;
      }
      res.add(prop);
    }
    return res;
  }
}


example/object_path_example.dart
import 'package:object_path/object_path.dart';
import 'package:pretty_json/pretty_json.dart';

void main() {
  //List Example
  var mock = <String, dynamic>{
    'Key 1': 'Val 1',
    'Key 2': 'Val 2',
    'Key 3': {'Key 3.1': 'Val 3'}
  };

  var objPath = ObjectPath();

  /*
  * Below we set three different keys on the map,
  * two of which are nested.
  * */

  mock = objPath.setList(mock, ['Key 4'], 'Val 4');
  mock = objPath.setList(mock, ['Key 5', 'Key 5.1'], 'Val 5.1');

  /*
  * Below we remove Key 2 and it's value from the map
  * */

  mock = objPath.unsetList(mock, ['Key 2']);

  /*
  * Here we expect an output of:
  * {
  *   'Key 1': 'Val 1',
  *   'Key 3': {
  *     'Key 3.1': 'Val 3'
  *   }
  *   'Key 4': 'Val 4',
  *   'Key 5':  {
  *     'Key 5.1': 'Val 5.1',
  *   }
  * }
  * */

  print(prettyJson(mock));

  //Dot Notation Example
  //reset the Map
  mock = <String, dynamic>{
    'Key 1': 'Val 1',
    'Key 2': 'Val 2',
    'Key 3': {'Key 3.1': 'Val 3'}
  };

  objPath = ObjectPath();

  /*
  * Below we set three different keys on the map,
  * two of which are nested.
  * */

  mock = objPath.setDot(mock, 'Key 4', 'Val 4');

  //Remember to escape unwanted splits using \\
  //Alternatively, you can specify your own split character using the
  //option 'splitAt' param
  mock = objPath.setDot(mock, 'Key 5.Key 5\\.1', 'Val 5.1');

  /*
  * Below we remove Key 2 and it's value from the map
  * */

  mock = objPath.unsetDot(mock, 'Key 2');

  /*
  * Here we expect an output of:
  * {
  *   'Key 1': 'Val 1',
  *   'Key 3': {
  *     'Key 3.1': 'Val 3'
  *   }
  *   'Key 4': 'Val 4',
  *   'Key 5':  {
  *     'Key 5.1': 'Val 5.1',
  *   }
  * }
  * */

  print(prettyJson(mock));

  print(objPath.getList(mock, ['Key 5','Key 5.1']));
  print(objPath.getDot(mock, 'Key 5.Key 5\\.1'));
}
